saleor/graphql/product/dataloaders/products.py

```diff
@@ -548,9 +547,13 @@ def batch_load(self, keys):
             product_collection_map[pid].append(cid)
 
         def map_collections(collections):
-            collection_map = {c.id: c for c in collections}
+            collection_map = {c.id: c for c in collections if c is not None}
             return [
-                [collection_map[cid] for cid in product_collection_map[pid]]
+                [
+                    collection_map[cid]
+                    for cid in product_collection_map[pid]
+                    if cid in collection_map
```

Comment:
@Air-t Dataloader should always return a list of the same length as the list of keys. This solution doesn't meet this requirement. I think you should start by creating a unit test with bug reproduction.

Reply:
I managed to reproduce it removing a collection while a dataloader `batch_load` was partially executed and `product_collection_pairs` list was created. Now both keys and return value lengths match. 
In a reproduced example, other functions had to be fixated to not fail while the collection was resolved as None by dataloader

Reply:
Could you write unit tests using the `before_after` module to reproduce the described case? We use this module in some tests e.g `saleor/checkout/tests/test_checkout_complete.py::test_complete_checkout_checkout_was_deleted_before_completing`